editor/plugins/curve_editor_plugin.cpp

```diff
@@ -33,276 +33,356 @@
+void CurveEdit::_notification(int p_what) {
 	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE:
+			connect("mouse_exited", callable_mp(this, &CurveEdit::_on_mouse_exited));
+			[[fallthrough]];
+		case NOTIFICATION_THEME_CHANGED:
+		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
+			float gizmo_scale = EDITOR_GET("interface/touchscreen/scale_gizmo_handles");
+			point_radius = Math::round(BASE_POINT_RADIUS * get_theme_default_base_scale() * gizmo_scale);
+			hover_radius = Math::round(BASE_HOVER_RADIUS * get_theme_default_base_scale() * gizmo_scale);
+			tangent_radius = Math::round(BASE_TANGENT_RADIUS * get_theme_default_base_scale() * gizmo_scale);
+			tangent_hover_radius = Math::round(BASE_TANGENT_HOVER_RADIUS * get_theme_default_base_scale() * gizmo_scale);
+			tangent_length = Math::round(BASE_TANGENT_LENGTH * get_theme_default_base_scale());
+		} break;
 		case NOTIFICATION_DRAW: {
 			_draw();
 		} break;
-		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
-			if (EditorSettings::get_singleton()->check_changed_settings_in_group("interface/touchscreen/scale_gizmo_handles")) {
-				_gizmo_handle_scale = EDITOR_GET("interface/touchscreen/scale_gizmo_handles");
+		case NOTIFICATION_VISIBILITY_CHANGED: {
+			if (!is_visible()) {
+				grabbing = GRAB_NONE;
 			}
 		} break;
 	}
 }
 
-void CurveEditor::gui_input(const Ref<InputEvent> &p_event) {
-	Ref<InputEventMouseButton> mb_ref = p_event;
-	if (mb_ref.is_valid()) {
-		const InputEventMouseButton &mb = **mb_ref;
+void CurveEdit::gui_input(const Ref<InputEvent> &p_event) {
+	ERR_FAIL_COND(p_event.is_null());
+
+	Ref<InputEventKey> k = p_event;
+	if (k.is_valid()) {
+		// Deleting points or making tangents linear.
+		if (k->is_pressed() && k->get_keycode() == Key::KEY_DELETE) {
+			if (selected_tangent_index != TANGENT_NONE) {
+				toggle_linear(selected_index, selected_tangent_index);
+			} else if (selected_index != -1) {
+				if (grabbing == GRAB_ADD) {
+					curve->remove_point(selected_index); // Point is temporary, so remove directly from curve.
+					set_selected_index(-1);
+				} else {
+					remove_point(selected_index);
+				}
+				grabbing = GRAB_NONE;
+				hovered_index = -1;
+				hovered_tangent_index = TANGENT_NONE;
+			}
+			accept_event();
+		}
 
-		if (mb.is_pressed() && !_dragging) {
-			Vector2 mpos = mb.get_position();
+		if (k->get_keycode() == Key::SHIFT || k->get_keycode() == Key::ALT) {
+			queue_redraw();
+		}
+	}
 
-			_selected_tangent = get_tangent_at(mpos);
-			if (_selected_tangent == TANGENT_NONE) {
-				set_selected_point(get_point_at(mpos));
+	Ref<InputEventMouseButton> mb = p_event;
+	if (mb.is_valid() && mb->is_pressed()) {
+		Vector2 mpos = mb->get_position();
+
+		if (mb->get_button_index() == MouseButton::RIGHT || mb->get_button_index() == MouseButton::MIDDLE) {
+			if (mb->get_button_index() == MouseButton::RIGHT) {
+				// Move a point to its old position.
+				if (grabbing == GRAB_MOVE) {
```

Comment:
`GRAB_MOVE` implies that you have grabbed an existing point and are moving it. In this case, RMB moves it back to its old position (like the 2D/3D editors); otherwise, the point gets deleted.